### ISSUE TITLE
Training ILSVRC12 on inception v4 / inception_resnet

### ISSUE THREAD

**Issue by revilokeb** [soeaver/caffe-model#1]
@soeaver have you by any chance done some training of ILSVRC12 on your inception v4 / inception_resnet2? I would be interested in an ILSVRC12 pre-trained model for both architectures, therefore I have started training inception_resnet2 a couple of days ago. But since this is taking quite some time I thought maybe you already have done some steps into that direction yourself and one could save some effort here / share?

**Comment by soeaver** [comment]
@revilokeb 
I'm sorry to tell you I don't have the pre-train models. 
To train the models for inception v4 or resnet2 will take a long time and must be very hard to close to the performance of papers.

**Comment by revilokeb** [comment]
@soeaver Thank you for your reply. 
By now I have  trained a network on ILSVRC12 (validation on all 50k) on your **inception_resnet_v2_train_test.prototxt** (only modifications: batch_size: 16 in phase: TRAIN and num_output: 1000 on layer "classifier") for 20 epochs.

Data is full ILSVRC12 train and val rescaled to 328x328. 

Solver is RMSprop with rms_decay: 0.9, delta: 1.0, gamma: 0.94, base_lr: 0.045 decreased every 2 epochs as in the publication. I did learning on 2 GPUs with iter_size: 2, so effective batch_size is 64.

In the publication http://arxiv.org/abs/1602.07261 Figure 24 shows a single crop Top-5 error on 50k validation set after 20 epochs of ~10% for inception-resnet-v2, Figure 26 a single crop Top-1 error of ~30%. 

In my own experiment results after 20 epochs are unfortunately much worse: single crop Top-5 validation error: 45%, Top-1: 70%, so I guess there is a major bug somewhere in my setup.

I will now go back and review the inception_resnet_v2_train_test.prototxt and the solver. 

I agree with you that coming up with a well trained model of similar accuracy as in the paper might be somewhat hard. But reaching similar accuracy after 20 epochs as stated in the paper should definitely not be impossible. If you (or anybody else) has got an idea as to the proper setup for training inception-resnet-v2 I am happy to hear.

**Comment by revilokeb** [comment]
I have restarted training of inception-resnet-v2 from scratch with small modifications of the solver: 1. reduced effective batch size to 32 and 2. removed small value of weight decay regularization which was present in my solver. The latter one seems to have a dramatic effect after first 2 epochs, i.e. now validation error seems to drop much quicker.

**Comment by soeaver** [comment]
@revilokeb 
I am so grateful to you for training inception-resnet-v2, and I think this is a very meaningful job for many people. In https://github.com/dmlc/mxnet-model-gallery/blob/master/imagenet-1k-inception-v3.md, maybe there is something useful to you, they crop 299 x 299 image from 384 x 384 image for training V3. This may improve some performance.
Now, my GPUs are used to train resnet-152 on ILSVRC2012, I'm very glad to share my work when the training is over. I also hope you can get a better performance on inception-resnet-v2.
Thank you again for your efforts!

**Comment by revilokeb** [comment]
@soeaver thanks, anyone interested might check https://github.com/revilokeb/inception_resnetv2_caffe where I am posting how training of that inception-resnet-v2 architecture is going. 

Would you mind me asking why you have decided to retrain resnet-152 on ILSVRC12 as this has been already released as pre-trained model in caffe (https://github.com/KaimingHe/deep-residual-networks)? I am sure it is a very valuable exercise but I assume also a lot of work and energy on your side...

**Comment by revilokeb** [comment]
@soeaver can I ask how your training of resnet-152 has been going and if you have reached similar perormance as He et al. or the FAIR guys?

With respect to your https://github.com/soeaver/caffe-model/blob/master/prototxts/inception_resnet_v2_train_test.prototxt, this has not been generated using your https://github.com/soeaver/caffe-model/blob/master/inception_resnet.py as the latter has got ReLUs after elementwise addition (residuals)

```
def eltwise_relu(bottom1, bottom2):
    residual_eltwise = L.Eltwise(bottom1, bottom2, eltwise_param=dict(operation=1))
    residual_eltwise_relu = L.ReLU(residual_eltwise, in_place=True)
```

whereas the former has not. I think the latter is more in line with the original Google publication.

Furthermore: 
- in https://github.com/soeaver/caffe-model/blob/master/inception_resnet.py  use_global_stats=False in BatchNorm layers is set but not lr_rate. Not 100% sure, but http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1BatchNormLayer.html#details seems to indicate that lr_rate must be set when using use_global_stats?
- in https://github.com/soeaver/caffe-model/blob/master/inception_resnet.py convolutional layers followed by BatchNorm do have a bias_term which seems to be superflous given the subsequent BatchNorm layer (however it might not do any harm).

**Comment by christopher5106** [comment]
Hi,
Do you have any news about these models ?
Thanks

**Comment by revilokeb** [comment]
@christopher5106 in case you are referring to inception-resnet-v2: I have run quite some number of experiments in addition (on Imagenet 299x299, 128x128 & CIFAR10; also made various adjustments) but so far have never been able to train my prototxt's of that architecture to any competitive accuracy. When wide residual nets came out I decided to give them I try and since then got stuck with them (https://github.com/revilokeb/wide_residual_nets_caffe). I might go back to inception-resnet-v2 when I have got some better intuition of wrn's. You might check https://github.com/lim0606/torch-inception-resnet-v2 for more succesfull training of inception-resnet-v2 and also a detailed account of their experiments in torch in the meantime.

**Comment by christopher5106** [comment]
@revilokeb thanks a lot. did you add any good news by chance on this topic ?

**Comment by ghost** [comment]
@soeaver   Where you able to train resnet-152 layer? If so can you share the Hyperparameter or prototxt files.   I am also trying the 152... But the accuracy remains flat...

**Comment by argman** [comment]
I have recently train a model on imagenet, here is my experience. you may get bad val accuracy because the use_global_stats set in bn layers, during the training, bn layers is unstable, so if you want to test your model during training, you should set use_global_stats=false, and when the training is almost over, you can set this back to true.

**Comment by christopher5106** [comment]
@argman what accuracy did you get on imagenet ?

**Comment by argman** [comment]
@christopher5106 I didnot train inception-v4, but another network.

**Comment by soeaver** [comment]
@tyrosathish  I am very sorry to tell you I didn't train resnet152 of my prototxt. If you want the pre-train models of resnets you can fork these works:
https://github.com/KaimingHe/deep-residual-networks
https://github.com/tornadomeet/ResNet
https://github.com/facebook/fb.resnet.torch
